libmcount/pyhook.c

```diff
@@ -0,0 +1,245 @@
+
+extern struct symtabs symtabs;
+
+enum py_args {
+	PY_ARG_ENTRY_ADDR = 0,
+	PY_ARG_RET_ADDR,
+	PY_ARG_SYMNAME,
+	PY_ARG_RETVAL,
+};
+
+/* The order has to be aligned with enum py_args */
+static const char *py_args_table[] = {
+	"entry_addr",
+	"ret_addr",
+	"symname",
+	"retval",
+};
+
+#define INIT_PY_API_FUNC(func) \
+	do { \
+		__##func = dlsym(python_handle, #func); \
+		if (!__##func) { \
+			pr_err("dlsym for \"" #func "\" is failed!\n"); \
+			python_enabled = false; \
+			return -1; \
+		} \
+	} while (0)
+
+static void remove_py_suffix(char *py_name)
+{
+	int len = strlen(py_name);
+	assert(len > 3 &&
+		py_name[len - 3] == '.' &&
+		py_name[len - 2] == 'p' &&
+		py_name[len - 1] == 'y');
+	py_name[len - 3] = '\0';
+}
+
+/* Import python module that is given by -p option */
+static int import_python_module(char *py_arg_pathname)
+{
+	char py_sysdir[PATH_MAX];
+	absolute_dirname(py_arg_pathname, py_sysdir);
+
+	/* Set path to import a python module. */
+	__PySys_SetPath(py_sysdir);
+	pr_dbg("PySys_SetPath(\"%s\") is done!\n", py_sysdir);
+
+	char *py_basename = basename(py_arg_pathname);
+	remove_py_suffix(py_basename);
+
+	pName = __PyString_FromString(py_basename);
+	pModule = __PyImport_Import(pName);
+	if (pModule == NULL) {
+		__PyErr_Print();
+		pr_warn("%s.py cannot be imported!\n", py_arg_pathname);
+		return -1;
+	}
+
+	return 0;
+}
+
+int python_uftrace_entry(unsigned long entry_addr, unsigned long ret_addr)
+{
+	PyObject *pythonArgument;
+	PyObject *pValue;
+
+	if (!pFuncEntry)
+		return -1;
+
+	struct sym *sym = find_symtabs(&symtabs, entry_addr);
+
+	/* Entire arguments are passed into a single dictionary. */
+	PyObject *pDict = __PyDict_New();
+
+	PyObject *pValue1 = __PyInt_FromLong(entry_addr);
+	PyObject *pValue2 = __PyInt_FromLong(ret_addr);
+	PyObject *pSym  = __PyString_FromString(sym->name);
+	assert(pDict && pValue1 && pValue2 && pSym);
+
+	__PyDict_SetItemString(pDict, py_args_table[PY_ARG_ENTRY_ADDR], pValue1);
+	__PyDict_SetItemString(pDict, py_args_table[PY_ARG_RET_ADDR], pValue2);
+	__PyDict_SetItemString(pDict, py_args_table[PY_ARG_SYMNAME], pSym);
+
+	/* Argument list must be passed in a tuple. */
+	pythonArgument = __PyTuple_New(1);
+	__PyTuple_SetItem(pythonArgument, 0, pDict);
+
+	pValue = __PyObject_CallObject(pFuncEntry, pythonArgument);
+
+	if (pValue != NULL)
+		pr_dbg("[python] uftrace_entry: %#x\n", __PyLong_AsLong(pValue));
+	else
+		__PyErr_Print();
+
+	return 0;
+}
+
+int python_uftrace_exit(unsigned long ret_addr, long *retval)
+{
+	PyObject *pythonArgument;
+	PyObject *pValue;
+
+	if (!pFuncExit)
+		return -1;
+
+	/* Entire arguments are passed into a single dictionary. */
+	PyObject *pDict = __PyDict_New();
+
+	PyObject *pValue1 = __PyInt_FromLong(ret_addr);
+	__PyDict_SetItemString(pDict, py_args_table[PY_ARG_RET_ADDR], pValue1);
+
+	if (retval) {
+		PyObject *pValue2 = __PyInt_FromLong(*retval);
+		__PyDict_SetItemString(pDict, py_args_table[PY_ARG_RETVAL], pValue2);
+	}
```

Comment:
I think the exit hook should provide the duration.  But it can be added later (with other useful fields too).